modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Producer.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2018-2023 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.storage.bigquery.streamloader
+
+import java.time.Instant
+import java.nio.charset.StandardCharsets
+
+import cats.implicits._
+
+import cats.effect.{Async, Resource}
+
+import com.permutive.pubsub.producer.{Model, PubsubProducer}
+import com.permutive.pubsub.producer.encoder.MessageEncoder
+import com.permutive.pubsub.producer.grpc.{GooglePubsubProducer, PubsubProducerConfig}
+
+import com.snowplowanalytics.snowplow.badrows.{BadRow, Failure, Payload => BadRowPayload}
+
+import org.typelevel.log4cats.Logger
+
+import scala.concurrent.duration._
+
+trait Producer[F[_], A] {
+  def produce(data: A): F[Unit]
+}
+
+object Producer {
+
+  val MaxPayloadLength = 9000000 // Stay under Pubsub Maximum of 10MB
+
+  def mkProducer[F[_]: Async: Logger, A: MessageEncoder](
+    projectId: String,
+    topic: String,
+    batchSize: Long,
+    delay: FiniteDuration,
+    oversizeBadRowProducer: Producer[F, BadRow.SizeViolation]
+  ): Resource[F, Producer[F, A]] =
+    mkProducer[F, A](projectId, topic, batchSize, delay).map { p =>
+      (data: A) => {
+        val dataSize = getSize(data)
+        if (dataSize >= MaxPayloadLength) {
+          val badRow = createSizeViolationBadRow(data, dataSize)
+          oversizeBadRowProducer.produce(badRow)
+        } else {
+          p.produce(data).void
+        }
+      }
+    }
+
+  def mkOversizeBadRowProducer[F[_]: Async: Logger](
+    projectId: String,
+    topic: String,
+    batchSize: Long,
+    delay: FiniteDuration
+  ): Resource[F, Producer[F, BadRow.SizeViolation]] =
+    mkProducer[F, BadRow](projectId, topic, batchSize, delay).map { p =>
+      (data: BadRow.SizeViolation) => p.produce(data).void
+    }
+
+  /** Construct a PubSub producer. */
+  private def mkProducer[F[_]: Async: Logger, A: MessageEncoder](
```

Comment:
Personally I try to avoid overloaded method names.  But I don't have a strong reason to complain about it.

Reply:
Yeah, I don't like that neither but since one of them is private, I thought it is okay.